pkg/virtual/generic/rootapiserver/root_apiserver.go

```diff
@@ -0,0 +1,369 @@
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+	cacheddiscovery "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/informers"
+	rbacinformers "k8s.io/client-go/informers/rbac/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	rbacregistryvalidation "k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbacauthorizer "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	virtualapiserver "github.com/kcp-dev/kcp/pkg/virtual/generic/apiserver"
+	"github.com/kcp-dev/kcp/pkg/virtual/generic/builders"
+)
+
+type InformerStart func(stopCh <-chan struct{})
+type InformerStarts []InformerStart
+
+type RootAPIExtraConfig struct {
+	builders.SharedExtraConfig
+
+	// we phrase it like this so we can build the post-start-hook, but no one can take more indirect dependencies on informers
+	informerStart func(stopCh <-chan struct{})
+
+	VirtualWorkspaces []builders.VirtualWorkspaceBuilder
+}
+
+// Validate helps ensure that we build this config correctly, because there are lots of bits to remember for now
+func (c *RootAPIExtraConfig) Validate() error {
+	ret := []error{}
+
+	if c.RuleResolver == nil {
+		ret = append(ret, fmt.Errorf("RuleResolver is required"))
+	}
+	if c.SubjectLocator == nil {
+		ret = append(ret, fmt.Errorf("SubjectLocator is required"))
+	}
+	if c.RESTMapper == nil {
+		ret = append(ret, fmt.Errorf("RESTMapper is required"))
+	}
+
+	return utilerrors.NewAggregate(ret)
+}
+
+type RootAPIConfig struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   RootAPIExtraConfig
+}
+
+// RootAPIServer is only responsible for serving the APIs for the virtual workspace
+// at a given root path or root path family
+// It does NOT expose oauth, related oauth endpoints, or any kube APIs.
+type RootAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	ExtraConfig   *RootAPIExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *RootAPIConfig) Complete() completedConfig {
+	cfg := completedConfig{
+		c.GenericConfig.Complete(),
+		&c.ExtraConfig,
+	}
+
+	return cfg
+}
+
+var _ builders.MainConfigProvider = (*completedConfig)(nil)
+
+func (c *completedConfig) CompletedConfig() genericapiserver.CompletedConfig {
+	return c.GenericConfig
+}
+
+func (c *completedConfig) SharedExtraConfig() builders.SharedExtraConfig {
+	return c.ExtraConfig.SharedExtraConfig
+}
```

Comment:
would drop the accessors. Just grab the struct directly.

Reply:
These are implemented methods of the `APIGroupConfigProvider`  interface that is in the `builders` package and that is used in a number of distinct packages.

Reply:
Not necessary anymore. This was removed.